src/datahike/db.cljc

```diff
@@ -1553,26 +1553,25 @@
         keep-history? (and (-keep-history? db) (not (no-history? db a-ident))
                            (not= :db/txInstant a-ident))
         op-count      (.op-count db)
-        old-val (when keep-history?
-                  (:v (first (di/-slice (:eavt db)
-                                        (dd/datom (.-e datom) (.-a datom) nil (.-tx datom))
-                                        (dd/datom (.-e datom) (.-a datom) nil (.-tx datom))
-                                        :eavt))))]
+        old-datom (first (di/-slice (:eavt db)
```

Comment:
Sorry, but just like `old-val` this doesn't belong in this place. As seen in the hitchhiker-tree, each index type has to decide for itself how to handle upserts, otherwise the index upsert function doesn't even make sense anymore. I will move this logic back into the index namespace. It is correct though that we can simply use the `:eavt` index to retrieve the old datom. I can try out if that is enough to have a decent transaction performance but I doubt it.